PoolSource TTreeCache refactoring

dan131riley opened the pull request:
#### PR description:

This PR moves all the TTreeCache management logic from the RootTree class into a new CacheManagerBase abstract base class and two specializations.  SparseReadCache implements the complicated cache-switching logic used by default.  SimpleCache implements a simplified version of the caching logic from #48088 that allows use of TTree::GetEntry.

Simplification of the simple caching case results in an improved IO profile.  By factoring out the two use cases, this also provides a better base for future modifications (especially to the simple caching case).

#### PR validation:

Tested on an AOD to MINIAOD production job similar to the one used for validation of #48088.  A limited matrix test job was also run.  This is a purely technical change to the ROOT file input system, no changes are expected in physics results.

resolves https://github.com/cms-sw/framework-team/issues/126

Commit:
initial checkin of cache manager

working refactoring of the standard read cache

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"156d3abc082e120d936616191eabcd9ca2c55e01":{"files":["FWCore/Utilities/interface/scope.h","IOPool/Input/src/InputFile.h","IOPool/Input/src/RootFile.cc","IOPool/Input/src/RootTree.cc","IOPool/Input/src/RootTreeCacheManager.cc","IOPool/Input/src/RootTreeCacheManager.h"],"squashed":false,"time":1758555114},"8946bf94de6797bc503c125985267edec6869102":{"files":["IOPool/Input/src/RootTreeCacheManager.cc","IOPool/Input/src/RootTreeCacheManager.h"],"squashed":false,"time":1758555114},"ac0f48dee765fbf3978503786bd62870e71227d5":{"files":["IOPool/Input/src/RootFile.cc","IOPool/Input/src/RootTree.cc","IOPool/Input/src/RootTree.h","IOPool/Input/src/RootTreeCacheManager.cc","IOPool/Input/src/RootTreeCacheManager.h"],"squashed":false,"time":1758555114},"cb2da01ef4b220aad9fd3caa3239ed0a7aeebe9f":{"files":["FWCore/Utilities/interface/scope.h","IOPool/Input/src/InputFile.h","IOPool/Input/src/RootFile.cc","IOPool/Input/src/RootTree.cc","IOPool/Input/src/RootTree.h","IOPool/Input/src/RootTreeCacheManager.cc","IOPool/Input/src/RootTreeCacheManager.h"],"squashed":false,"time":1758555113}},"emoji":{},"last_seen_sha":"8946bf94de6797bc503c125985267edec6869102","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48969/46138

- Found files with invalid states:
  - FWCore/Utilities/interface/scope.h:
    - Added: cb2da01ef4b220aad9fd3caa3239ed0a7aeebe9f
    - Deleted: 156d3abc082e120d936616191eabcd9ca2c55e01

dan131riley commented:
Limited matrix found a problem, converted to draft until debugged.

Commit:
Always use the simple cache w/o the prompt delayed reader for
non-event trees.

Add a policy that does not cache at all, and a policy that adds
a separate EventAuxiliary cache to the simple policy.

cmsbuild commented:
Pull request #48969 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

makortel commented on IOPool/Input/src/RootTreeCacheManager.cc:
```diff
@@ -0,0 +1,544 @@
+        // ROOT will automatically expand the cache to fit one cluster; hence, we use
+        // 5 MB as the cache size below
+        rawTriggerTreeCache_ = createCacheWithSize(5 * 1024 * 1024);
+        if (rawTriggerTreeCache_) {
+          rawTriggerTreeCache_->SetEnablePrefetching(false);
+          TObjArray* branches = tree_->GetListOfBranches();
+          int branchCount = branches->GetEntriesFast();
+
+          // Train the rawTriggerCache to have everything not in the regular cache.
+          rawTriggerTreeCache_->SetLearnEntries(0);
+          rawTriggerTreeCache_->SetEntryRange(entryNumber, rawTriggerSwitchOverEntry_);
+          for (int i = 0; i < branchCount; i++) {
+            auto tmp_branch = dynamic_cast<TBranch*>(branches->UncheckedAt(i));
+            if (trainedSet_.find(tmp_branch) != trainedSet_.end()) {
+              continue;
+            }
+            rawTriggerTreeCache_->AddBranch(tmp_branch, kTRUE);
+          }
+          rawTriggerTreeCache_->StopLearningPhase();
+        }
+        performedSwitchOver_ = false;
```
Just to note that in the earlier code the `if` behavior was different
https://github.com/cms-sw/cmssw/blob/f934747fc3f3b0c983bafb0b6e8a8571c80f5f3a/IOPool/Input/src/RootTree.cc#L347-L363
but given that `rawTriggerTreeCache_` was used unconditionally after the first `if` maybe that was a bug and the structure here looks like plausible intended behavior.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48969/46234

- Found files with invalid states:
  - FWCore/Utilities/interface/scope.h:
    - Added: cb2da01ef4b220aad9fd3caa3239ed0a7aeebe9f
    - Deleted: 156d3abc082e120d936616191eabcd9ca2c55e01

cmsbuild commented:
Pull request #48969 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

dan131riley commented on IOPool/Input/src/RootTreeCacheManager.h:
```diff
@@ -0,0 +1,62 @@
+      virtual ~CacheManagerBase() = default;
+
+      virtual void createPrimaryCache(unsigned int cacheSize) = 0;
+      virtual void setEntryNumber(EntryNumber theEntryNumber, EntryNumber entryNumber, EntryNumber entries) = 0;
```
Renamed to nextEntryNumber (I never really like theEntryNumber) and a comment added

dan131riley commented on IOPool/Input/src/RootTreeCacheManager.h:
```diff
@@ -0,0 +1,62 @@
+      virtual void getAuxEntry(TBranch* auxBranch, EntryNumber entryNumber);
+      virtual void getEntryForAllBranches(EntryNumber entryNumber) const = 0;
+
+      static std::unique_ptr<CacheManagerBase> create(const std::string& strategy,
```
done

makortel commented:
@cmsbuild, please test

dan131riley commented on FWCore/Framework/bin/BuildFile.xml:
```diff
@@ -13,7 +13,8 @@
 
-<bin name="cmsRunGlibC" file="cmsRun.cpp">
+<bin name="cmsRunGlibC" file="cmsRun.cpp *.cxx">
+  <use name="rootimt"/>
```
No, leftovers from some root debugging, will revert

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48969/46235

- Found files with invalid states:
  - FWCore/Utilities/interface/scope.h:
    - Added: cb2da01ef4b220aad9fd3caa3239ed0a7aeebe9f
    - Deleted: 156d3abc082e120d936616191eabcd9ca2c55e01

cmsbuild commented:
Pull request #48969 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48969/46270

makortel commented:
+core

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

Commit:
simple cache takes ownership of its cache
fix training of simple cache when promptRead is true
add NoCache and SimpleWithAuxCache

makortel commented on IOPool/Input/src/RootTreeCacheManager.cc:
```diff
@@ -0,0 +1,601 @@
+      if (cachestats) {
+        treeCache_->Print("a cachedbranches");
+      }
+      //treeCache_->DropBranch("*");
```
Is this commented-out line still useful?

makortel commented:
Just for me to understand, is the `SimpleWithAuxCache` perhaps something that you tried but didn't work as well as the `SimpleCache`?

Commit:
correct two rebasing errors

cmsbuild commented:
Pull request #48969 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

makortel commented:
enable threading

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (test failures were overridden). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @mandrenguyen, @ftenchini (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
